lib/AST/ModuleNameLookup.cpp

```diff
@@ -90,20 +95,111 @@ static bool updateOverloadSet(NamedCanTypeSet &overloads,
   return true;
 }
 
-/// After finding decls by name lookup, filter based on the given
-/// resolution kind and existing overload set and add them to \p results.
+/// Returns whether it's okay to stop recursively searching imports, given that
+/// we found something, are using a CanTypeSet to resolve overloads, and have
+/// the given \p resolutionKind.
+static bool canReturnEarly(const CanTypeSet &overloads,
+                           ResolutionKind resolutionKind) {
+  switch (resolutionKind) {
+  case ResolutionKind::Overloadable:
+    return false;
+  case ResolutionKind::Exact:
+  case ResolutionKind::TypesOnly:
+    return true;
+  }
+}
+
+/// Returns whether it's okay to stop recursively searching imports, given that
+/// we found something, are using a NamedCanTypeSet to resolve overloads, and
+/// have the given \p resolutionKind.
+static bool canReturnEarly(const NamedCanTypeSet &overloads,
+                           ResolutionKind resolutionKind) {
+  return false;
+}
+
+namespace {
+using LookupCallbackTy = llvm::function_ref<
+    void(ModuleDecl *, ModuleDecl::AccessPathTy,
+         SmallVectorImpl<ValueDecl *> &)>;
+
+/// Encapsulates the work done for a recursive qualified lookup into a module.
+///
+/// The \c OverloadSetTy chosen determines how overloads are resolved.
 template <typename OverloadSetTy>
-static ResolutionKind recordImportDecls(LazyResolver *typeResolver,
-                                        SmallVectorImpl<ValueDecl *> &results,
-                                        ArrayRef<ValueDecl *> newDecls,
-                                        OverloadSetTy &overloads,
-                                        ResolutionKind resolutionKind) {
+class ModuleNameLookup {
+  ModuleLookupCache cache;
+  LookupCallbackTy callback;
+  LazyResolver *typeResolver;
+  bool respectAccessControl;
+
+  ModuleNameLookup() = default;
+
+  /// After finding decls by name lookup, filter based on the given
+  /// resolution kind and existing overload set and add them to \p results.
+  ///
+  /// Both \p overloads and \p resolutionKind are updated based on the new
+  /// declarations.
+  ///
+  /// \returns true if any decls were added to \p results.
+  bool recordImportDecls(SmallVectorImpl<ValueDecl *> &results,
+                         ArrayRef<ValueDecl *> newDecls,
+                         OverloadSetTy &overloads,
+                         ResolutionKind &resolutionKind);
```

Comment:
Is resolutionKind just an input parameter? If so, could it be `const`?

Reply:
No, it's in/out. I may not use `const` on pass-by-value parameters, but I always do it for references.